### ISSUE TITLE
Inaccurate test results 

### ISSUE THREAD

**Issue by OmarTawfik** [actions/workflow-parser#42]
I started using running the tests from this repository in my [JS parser](https://github.com/OmarTawfik/github-actions-js/issues/79). I found a number of issues where my parser generates more actions than the the go parser, as it recovers from syntax errors.

Example from `tests/invalid/bad-hcl-4` (my parser produces `1` action instead of `0`):

```
# Invalid file, because a string is not terminated.

action "foo" { uses=" }

# ASSERT {
#   "result":       "failure",
#   "numActions":   0,
#   "numWorkflows": 0,
#   "errors":[
#     { "severity": "FATAL", "message": "literal not terminated" }
#   ]
# }
```

Is this something that is currently WIP in the go parser? Recovering from syntax errors is crucial for IDE support. Any thoughts on how to resolve this?

**Comment by OmarTawfik** [comment]
Here is a list of the failing tests, with their *corrected* number of actions:
```
  ["bad-hcl-2.workflow", 1],
  ["bad-hcl-3.workflow", 1],
  ["bad-hcl-4.workflow", 1],
  ["bad-hcl-5.workflow", 1],
  ["hcl-subset.workflow", 4],
```

**Comment by OmarTawfik** [comment]
cc @piki

**Comment by piki** [comment]
Good point, @OmarTawfik.

`bad-hcl-*.workflow` is pretty much guaranteed to return zero actions, as long as the Go parser relies on the `hashicorp/hcl` parser to extract the AST from the string, because the `hashicorp/hcl` parser doesn't recover from errors at all.  All our current error recovery is at the semantic level, once the AST is complete.  I think that's a likely limitation for the foreseeable future, because we don't plan to switch away from `hashicorp/hcl`, nor do we plan to add AST-level error handling to it.

Error recovery is effectively an additional syntax to parse that is defined only in the parser code, not in any formal grammar.  Without a shared grammar to target, it's hard to make the Go code and the JS code recover errors in exactly the same way.  To wit: we both consider `action "a" { uses={} }` to produce an action, and we both consider `akshun "a" { ... }` does not, but we have different results for `action "a" { uses="x }`.

One fuzzy approach would be for the JSON expectations in the test fixtures to accept `minNumWorkflows` and `maxNumWorkflows` (i.e., a range) in place of a set `numWorkflows`, for cases like [the five you identified](https://github.com/actions/workflow-parser/issues/42#issuecomment-475910573) where there's some ambiguity about how aggressively to recover from errors.

What do you think?

**Comment by OmarTawfik** [comment]
Since that is not a goal for this project, I'll keep a map of actual vs recovered counts, and assert as such in my tests. Thanks for clarifying!
